FOCUS #149: CommitmentDiscountId spec changes and supporting content (#166)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/dimensions/commitmentid.md

> # Commitment ID

Reviewer: Can we change this to "Commitment Discount ID"? (Same for the name column.)

Author: You mean any references to "Commitment" should be changed to "Commitment Discount", right? (include filenames as well: `commitmentdiscountid.md`, etc.)

Author: Renamed.

> A commitment is a contractual commitment to use a minimum amount of spending on resources for a fixed term in exchange for discounts on those resources. Commitments can cover various resources such as virtual machines, or databases. A provider can offer various pricing models, payment options, and term durations for its commitment offerings.

Reviewer: A few thoughts:
1. Thinking out loud: Should we say "A commitment-based discount is..."? There are other commitments that aren't "commitment-based discounts". Not sure if we want to generalize this to include those or not.
2. Does "minimum" make it sound like they're only valuable if you have 100% utilization?
3. We should work in a reference to differentiate usage-based and cost-based commitments. (Added something, but I feel like it could be better.)
```suggestion
A commitment-based discount is a contractual commitment to use a certain amount of a specific type of usage for a fixed term in exchange for a discounted unit price. Commitment-based discounts can cover either usage or cost of various resources such as virtual machines or databases. A provider can offer various pricing models, payment options, and term durations for commitment-based discounts.
```

> A commitment-based discount is a contractual commitment to use a certain amount of a specific type of usage for a fixed term in exchange for a discounted unit price. Commitment-based discounts can cover either usage or cost of various resources such as virtual machines or databases. A provider can offer various pricing models, payment options, and term durations for commitment-based discounts.

Reviewer 2: I really like this paragraph, but it is nothing specifically to do with the ID. ie. it applies to all Commitment fields - it just happens to be on the top of this one to give it context. Does this mean all Commitment Discount fields should have it? Or should it be abstracted into a base class?

Author: There's [this](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/glossary.md) file. It's empty though.

Author: Pretty much patterned with BillingAccount[Id|Name], Resource[Id|Name], and SubAccount[Id|Name] columns.

Reviewer: We'll mention this to Karl as part of the glossary effort! We'll clean this up as needed once we have the right guidance. Closing this for now.

> A Commitment ID is the identifier assigned to a commitment by the provider.

Reviewer: ```suggestion
A Commitment ID is the identifier assigned to a commitment-based discount by the provider.
```

> A Commitment ID is the identifier assigned to a commitment-based discount by the provider.

Reviewer: Mention the scenarios that the column can be used to address, like commitment utilization and savings.

Reviewer: Made an explicit suggestion to address this feedback in the latest iteration. Closing.

> The CommitmentId column MUST be present in the billing data. This column must be of type String and MUST NOT contain null values. CommitmentId MUST be a globally unique identifier within a provider.

Reviewer: Saying "globally unique identifier" sounds like we require it to be a GUID. 
```suggestion
The CommitmentId column MUST be present in the billing data. This column must be of type String and MUST NOT contain null values when a charge is related to a commitment-based discount. When a charge is not associated with a commitment-based discount, the column MUST be null. CommitmentId MUST be unique within the provider.
```

## Column ID

CommitmentId

## Display name

Commitment ID

## Description

> The identifier assigned to a commitment by the provider.

Reviewer: ```suggestion
The identifier assigned to a commitment-based discount by the provider.
```

> Commitment ID

Reviewer 2: Just double-checking that we're 100% sure this should not be Id?

## Content constraints

|    Constraint   |      Value      |
|:----------------|:----------------|
| Column required | True            |
| Data type       | String          |

> | Allows nulls    | False            |

Reviewer: Need to support nulls for charges that aren't related to CBDs.
```suggestion
| Allows nulls    | True             |
```
